src/models/Throttler.ts

```diff
@@ -1,45 +1,191 @@
+    /**
+     * Resets the throttler.
+     * @param id
+     */
+    reset(id: string): void;
+    /**
+     * Increment the throttler's counter.
+     * @param message
+     * @returns Either or not the limit has been reached.
+     * @example
+     * ```ts
+     * if (throttler.add(message)) {
+     *     // Do action
+     * } else {
+     *     // Cannot do action
+     * }
+     * ```
+     */
+    add(message: Message): boolean;
+}
+
+export class CommandThrottler {
+    /**
+     * @param scope The scope of the throttler.
      * @param count How many time the throttler can be triggered.
      * @param duration Time in seconds since the first trigger before the throttler is reset.
      */
+    constructor(scope: ThrottlerScope, count: number, duration: number) {
+        if (scope === "global") return new Throttler(count, duration);
+        else return new ScopedThrottler(scope, count, duration);
+    }
+}
+
+class Throttler implements CommandThrottler {
+    public readonly scope: "global" = "global";
+    private _current = 0;
+    private _timeout: NodeJS.Timeout | undefined = undefined;
+
     constructor(public readonly count: number, public readonly duration: number) {}
 
-    /** The number of time this throttler has been triggered. */
-    get current() {
+    getCurrent(): number {
         return this._current;
     }
-
-    /** Either or not this throttler has reached the limit. */
-    get throttled() {
+    getThrottled(): boolean {
         return this._current >= this.count;
     }
-
-    /** The time in seconds until the throttler is reset. */
-    get cooldown() {
+    getCooldown(): number {
         return this._timeout ? Math.ceil(getTimeLeft(this._timeout)) : 0;
     }
 
-    /** Resets this throttler. */
     reset(): void {
         if (this._timeout) clearTimeout(this._timeout);
         this._timeout = undefined;
         this._current = 0;
     }
 
-    /**
-     * Increment the counter.
-     * @returns Either or not the limit has been reached.
-     */
     add(): boolean {
-        const reachedLimit = this.throttled;
+        const reachedLimit = this.getThrottled();
         this._current++;
         if (!this._timeout) this._timeout = setTimeout(() => this.reset(), this.duration * 1000);
         return reachedLimit;
     }
 }
+
+interface ThrottlerData {
+    current: number;
+    timeout: NodeJS.Timeout;
+}
+
+class ScopedThrottler implements CommandThrottler {
+    private throttlers = new Map<string, ThrottlerData>();
+
+    constructor(
+        public readonly scope: Exclude<ThrottlerScope, "global">,
+        public readonly count: number,
+        public readonly duration: number,
+    ) {}
+
+    private getKey(message: Message) {
+        switch (this.scope) {
+            case "user":
+                return message.author.id;
+            case "member":
+                return message.member?.id ?? "GLOBAL";
+            case "guild":
+                return message.guild?.id ?? "GLOBAL";
```

Comment:
What's happen if scope is `member` or `guild` and the message don't come from a guild ?

**example**: the message is a DM.

**possible solutions**:
- Don't throttle, *but the user can spam the command*.
- Only allow `member` and `guild` scope for guild-only command.
- Use the `user.id` instead.
- Use a global key.